diff --git a/exp_run.sh b/exp_run.sh
new file mode 100644
index 00000000..9d058795
--- /dev/null
+++ b/exp_run.sh
@@ -0,0 +1,38 @@
+#!/bin/bash
+
+number_of_agents=15
+number_of_jobs=$((2 * number_of_agents))
+
+# seconds
+time=86400
+
+working_dir=$(pwd)
+
+method="metamut_i"
+
+seeds=(2034567890 157839206 4037289401 1324675832 2947583017 1908234765 3284957103 2763891275 3746283957 1092387456)
+
+# Loop from 0 to 4
+for i in {0..4}; do
+    dir="$method-$i"
+    
+    # Create the directory if it doesn't exist
+    mkdir -p "$dir"
+
+    # Navigate into the directory
+    cd "$dir" || exit
+
+    echo ${seeds[$i]}
+
+    # Run the Python script
+    python3 ../fuzzer/run.py -j ${number_of_jobs} \
+      --wdir $working_dir/$dir \
+      --repeat-times ${number_of_agents} \
+      --duration ${time} \
+      --seeds-dir $working_dir/seeds \
+      --cc-opt=-O2 \
+      --seed ${seeds[$i]}
+
+    # Return to the original directory
+    cd ..
+done
diff --git a/fuzzer/configs.py b/fuzzer/configs.py
index 36cac5b3..5338e290 100644
--- a/fuzzer/configs.py
+++ b/fuzzer/configs.py
@@ -5,7 +5,7 @@ import pathlib
 homedir = pathlib.Path(__file__).parent.parent
 
 muss_homedir = f"{homedir}/mutators"
-muss_bin = f"{muss_homedir}/output/muss"
+muss_bin = f"{muss_homedir}/output/metamut_i"
 
 seeds_dir = f"{homedir}/seeds"
 
@@ -17,11 +17,11 @@ clang_bin = f"{llvm_objdir}/bin/clang"
 
 instrum_gcc_objdir = f"{homedir}/objects/gcc-project-trunk.instrum"
 instrum_gcc_bin = f"{instrum_gcc_objdir}/install/bin/gcc"
-instrum_gcc_nbrs = 1478061
+instrum_gcc_nbrs = 1573698
 
 instrum_llvm_objdir = f"{homedir}/objects/llvm-project-trunk.instrum"
 instrum_clang_bin = f"{instrum_llvm_objdir}/bin/clang"
-instrum_clang_nbrs = 4915140
+instrum_clang_nbrs = 4422726
 
 cc_map = {
     gcc_bin: instrum_gcc_bin,
diff --git a/fuzzer/lib/BugInst.py b/fuzzer/lib/BugInst.py
index d3af9a40..54f37cf4 100644
--- a/fuzzer/lib/BugInst.py
+++ b/fuzzer/lib/BugInst.py
@@ -218,6 +218,11 @@ class FromBugTracks:
       "__GI_abort",
       "__assert_fail_base",
       "__GI___assert_fail",
+      #### added ####
+      "diagnostic_context::action_after_output",
+      "diagnostic_context::report_diagnostic",
+      "diagnostic_context::diagnostic_impl",
+      ###############
       "diagnostic_action_after_output",
       "diagnostic_report_diagnostic",
       "diagnostic_impl",
@@ -233,7 +238,19 @@ class FromBugTracks:
       "llvm::report_fatal_error(char const*, bool)",
       "llvm::llvm_unreachable_internal(char const*, char const*, unsigned int)",
       "tree_class_check",
-      "llvm::CastInst::Create(llvm::Instruction::CastOps, llvm::Value*, llvm::Type*, llvm::Twine const&, llvm::Instruction*)",
+      #### added ####
+      # "tree_class_check_failed",
+      "tree_check",
+      # "tree_check_failed",
+      "tree_not_check",
+      # "tree_not_check_failed",
+      "tree_ssa_strip_useless_type_conversions",
+      "gimplify_expr",
+      "c_gimplify_expr",
+      "_fatal_insn",
+      "_fatal_insn_not_found",
+      ###############
+      "llvm::CastInst::Create(llvm::Instruction::CastOps, llvm::Value*, llvm::Type*, llvm::Twine const&, llvm::InsertPosition)",
       "llvm::IRBuilderBase::CreateCast(llvm::Instruction::CastOps, llvm::Value*, llvm::Type*, llvm::Twine const&)",
     ]
   @staticmethod
diff --git a/fuzzer/lib/ControlledProcess.py b/fuzzer/lib/ControlledProcess.py
index 9d51c480..b22409cf 100644
--- a/fuzzer/lib/ControlledProcess.py
+++ b/fuzzer/lib/ControlledProcess.py
@@ -230,7 +230,8 @@ class ControlledGDB:
     last_time = time.time()
     counter = {}
     while time.time() < last_time + duration:
-      self.run(cmd='c', timeout=step)
+      if not self.run(cmd='c', timeout=step):
+        break
       response = self.communicate('-stack-list-frames 0 {depth}')
       funcs = re.findall(r'func="[^"]*"', response)
       for func in funcs:
@@ -280,3 +281,4 @@ class ControlledGDB:
     time.sleep(0.1)
     self.p.terminate()
     self.p.wait()
+    self.pmon.kill_procs()
diff --git a/fuzzer/lib/Fuzzer.py b/fuzzer/lib/Fuzzer.py
index 1d00a69e..4d4abb12 100644
--- a/fuzzer/lib/Fuzzer.py
+++ b/fuzzer/lib/Fuzzer.py
@@ -52,6 +52,8 @@ class CompilerTracer:
     self.gdb.communicate('b llvm::sys::PrintStackTrace')
     # True means normally exits, False means timeout
     res = self.gdb.run(timeout=self.timeout)
+    if not res:
+      return (res, None)
     # return (res, self.gdb.traceback_functions())
     # return (res, self.gdb.traceback_functions_and_addresses())
     return (res, self.gdb.traceback_functions_and_lines())
@@ -184,7 +186,12 @@ class Fuzzer:
           j = [v.to_json() for v in self.crashes.values()]
           fp.write(json.dumps(j, indent=2))
     return True
-
+  
+  def save_crash(self, cc, testcase):
+    with open(f"{self.fuzz_args.wdir}/crash.log", 'a+') as fp:
+      now = time.time() - self.start_time
+      fp.write(f'{now}: [{cc}] Found bug report message in {testcase}\n')
+    
   def fuzz(self):
     if self.query_current_state() == 'done':
       wlog('done detected, quit now')
@@ -200,30 +207,50 @@ class Fuzzer:
       opfx = self.next_opfx()
       ot = self.fuzz_args.mutator.mutate(ifile, opfx)
       if not ot: continue
-      if self.fuzz_args.coverage_only:
-        has_crash = False
+      
+      has_crash = False
+      # compile with GCC
+      if self.fuzz_args.cc.name == 'gcc':
+        prev_coverage = self.sampler.sample()
+        with self.sampler:
+          for opt_str in configs.options:
+            pRes = self.compile(configs.instrum_gcc_bin, ot, options=shlex.split(opt_str), capture_stderr=True)
+        curr_coverage = self.sampler.sample()
+      else:
+        for opt_str in configs.options:
+          pRes = self.compile(configs.gcc_bin, ot, options=shlex.split(opt_str), capture_stderr=True)
+      
+        chkRes = CompilerStatus.analyze(pRes.retcode, pRes.stderr, pRes.is_tle, pRes.is_oom)
+        if chkRes == CompilerStatus.Crash:
+          self.save_crash("GCC",ot)
+          has_crash = True
+      
+      # compile with Clang
+      if self.fuzz_args.cc.name == 'clang':
+        prev_coverage = self.sampler.sample()
+        with self.sampler:
+          for opt_str in configs.options:
+            pRes = self.compile(configs.instrum_clang_bin, ot, options=shlex.split(opt_str), capture_stderr=True)
+        curr_coverage = self.sampler.sample()
       else:
-        has_crash = self.check_crash(ot)
-      # check coverage
-      prev_coverage = self.sampler.sample()
-      with self.sampler:
         for opt_str in configs.options:
-          self.compile(self.fuzz_args.cc.instrum_path,
-              ot, options=shlex.split(opt_str))
-      curr_coverage = self.sampler.sample()
+          pRes = self.compile(configs.clang_bin, ot, options=shlex.split(opt_str), capture_stderr=True)
+      
+        chkRes = CompilerStatus.analyze(pRes.retcode, pRes.stderr, pRes.is_tle, pRes.is_oom)
+        if chkRes == CompilerStatus.Crash:
+          self.save_crash("Clang",ot)
+          has_crash = True
+      
       if curr_coverage.outperforms(prev_coverage):
         # wlog(f"new branch: {ot}")
         self.save_coverage(curr_coverage)
-        if self.fuzz_args.mutator.is_generator():
-          if not has_crash:
-            self.fuzz_args.mutator.clean(ot)
-        else:
-          assert len(ot.ifiles) == 1
+        assert len(ot.ifiles) == 1
+        ## added a condition (if not has_crash)
+        if not has_crash:
           self.fuzz_args.testcases.append(ot.ifiles[0])
         self.last_save_time = time.time()
       else:
         if not has_crash:
-          # wlog(f"bored testcase: {ot}")
           self.fuzz_args.mutator.clean(ot)
         if time.time() >= self.last_save_time + 600:
           self.last_save_time = time.time()
@@ -231,3 +258,11 @@ class Fuzzer:
       # clean tmpdir
       Utils.remove_files(self.tmpdir)
     self.mark_current_state('done')
+
+    with open(f"{self.fuzz_args.wdir}/seeds.txt", 'w') as fp:
+      for testcase in self.fuzz_args.testcases:
+        base_filename = os.path.basename(testcase)
+        if re.match(r'^\d+\.c$', base_filename):
+          continue
+        else:
+          fp.write(testcase + "\n")
diff --git a/fuzzer/lib/Muss.py b/fuzzer/lib/Muss.py
index 7279c5b5..e59305dd 100644
--- a/fuzzer/lib/Muss.py
+++ b/fuzzer/lib/Muss.py
@@ -49,6 +49,7 @@ class Muss(Mutator):
     seed = random.randint(0, 2**31 - 1)
     cmd  = f'{self.mutator_bin} --seed {seed}'
     cmd += f' --randomly-try-all-mutators'
+    cmd += f' --print-mutations'
     cmd += f' -i {ifile} -o {ofile}'
     self.invoke_cmd = cmd
     p = ControlledProcess(cmd,
diff --git a/fuzzer/lib/Schedule.py b/fuzzer/lib/Schedule.py
index bdd039d8..f70982b8 100644
--- a/fuzzer/lib/Schedule.py
+++ b/fuzzer/lib/Schedule.py
@@ -8,7 +8,8 @@ class Scheduler:
     self.worker_args = worker_args
     self.jobs = jobs
     self.ps = []
-    self.cores = list(range(self.jobs))
+    # self.cores = list(range(self.jobs))
+    self.cores = psutil.Process().cpu_affinity()
 
   def execute_jobs(self):
     jobs = iter(self.worker_args)
@@ -28,6 +29,6 @@ class Scheduler:
       p = Process(target=self.worker, args=[job])
       p.start()
       self.ps.append([affinity, p, job])
-      psutil.Process(p.pid).cpu_affinity([
-        affinity, affinity + psutil.cpu_count() // 2])
+      # psutil.Process(p.pid).cpu_affinity()
+      psutil.Process(p.pid).cpu_affinity([affinity])
     for _, p, _ in self.ps: p.join()
diff --git a/fuzzer/show.py b/fuzzer/show.py
index 13cb76d4..0ce2e504 100644
--- a/fuzzer/show.py
+++ b/fuzzer/show.py
@@ -36,9 +36,9 @@ def load_bugs(dirs):
   bugmap = {}
   for d in dirs:
     d = pathlib.Path(d)
-    st_date = parse_start_date(d/'log.txt')
     if not d.is_dir(): continue
     if ',' not in str(d): continue
+    st_date = parse_start_date(d/'log.txt')
     mutator, cc, num = str(d).split(',')
     jfile = d/'crashes.json'
     if os.access(jfile, os.R_OK):
diff --git a/mutators/CMakeLists.txt b/mutators/CMakeLists.txt
index 20d917f3..9b7a8a82 100644
--- a/mutators/CMakeLists.txt
+++ b/mutators/CMakeLists.txt
@@ -3,6 +3,9 @@ project(MUSS CXX C)
 include(cmake/string_to_list.cmake)
 include(cmake/find_llvm.cmake)
 
+# set(CMAKE_BUILD_TYPE Debug)
+# set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -O0 -g")
+
 set(CMAKE_C_COMPILER "gcc")
 set(CMAKE_CXX_COMPILER "g++")
 
@@ -47,9 +50,19 @@ function(add_muss_executable name)
   endif()
 endfunction()
 
-add_muss_executable(muss
+add_muss_executable(metamut
   FILES main.cpp "mu*/*.cpp" "mu*/*.cpp")
 add_muss_executable(mu.s
   FILES main.cpp "mu.s/*.cpp" "mu.s/*.cpp")
 add_muss_executable(mu.u
   FILES main.cpp "mu.u/*.cpp" "mu.u/*.cpp")
+add_muss_executable(bhis
+  FILES main.cpp "bhis/*.cpp" "bhis/*.cpp")
+add_muss_executable(bhis-gcc
+  FILES main.cpp "bhis_gcc_sample/*.cpp" "bhis_gcc_sample/*.cpp")
+add_muss_executable(bhis-llvm
+  FILES main.cpp "bhis_llvm_sample/*.cpp" "bhis_llvm_sample/*.cpp")
+add_muss_executable(metamut_i
+  FILES main.cpp "metamut_i/*.cpp" "metamut_i/*.cpp")
+add_muss_executable(successful
+  FILES main.cpp "successful/*.cpp" "successful/*.cpp")
diff --git a/mutators/include/Mutator.h b/mutators/include/Mutator.h
index 8301173c..543f5b3d 100644
--- a/mutators/include/Mutator.h
+++ b/mutators/include/Mutator.h
@@ -103,6 +103,7 @@ public:
 
   clang::ASTContext &getASTContext() const;
   clang::CompilerInstance &getCompilerInstance() const;
+  unsigned getSeed() const;
   MutationRange getMutationRange() const;
 
   const char *getName() const { return name; }
diff --git a/mutators/include/MutatorManager.h b/mutators/include/MutatorManager.h
index 6d2b34e0..1c56fa76 100644
--- a/mutators/include/MutatorManager.h
+++ b/mutators/include/MutatorManager.h
@@ -64,6 +64,8 @@ public:
     return mutationRange;
   }
 
+  unsigned getSeed() const { return SEED; }
+
   llvm::raw_ostream &outs();
 
   MutatorManager() = default;
@@ -72,6 +74,7 @@ private:
   std::string ErrorMsg;
 
   unsigned seed = 0;
+  unsigned SEED = 0;
   std::unique_ptr<std::mt19937> rndgen;
   std::string mutatorName;
   std::string inputFile;
diff --git a/mutators/include/MutatorUsingBash.h b/mutators/include/MutatorUsingBash.h
new file mode 100644
index 00000000..90a3b10a
--- /dev/null
+++ b/mutators/include/MutatorUsingBash.h
@@ -0,0 +1,58 @@
+#ifndef MUTATOR_USING_BASH_H
+#define MUTATOR_USING_BASH_H
+
+#include <clang/AST/ASTContext.h>
+#include <clang/AST/RecursiveASTVisitor.h>
+#include <clang/Basic/SourceManager.h>
+#include <clang/Lex/Lexer.h>
+#include <string>
+#include <array>
+#include <iostream>
+#include "Mutator.h"
+
+using namespace clang;
+
+class MutatorUsingBash : public Mutator {
+public:
+    using Mutator::Mutator;
+
+    MutatorUsingBash(const char* name, const char* desc)
+        : Mutator(name, desc) {
+        // Additional initialization (if needed)
+    }
+
+    /**
+     * Replaces all occurrences of a pattern in the input string with a replacement string.
+     * @param pattern The pattern to replace.
+     * @param replacement The replacement string.
+     * @return True if the replacement was successful, false otherwise.
+     */
+    bool replaceAll(std::string pattern, std::string replacement);
+
+    /**
+     * Replaces the random occurrence of a pattern in the input string with a replacement string.
+     * @param pattern The pattern to replace.
+     * @param replacement The replacement string.
+     * @return True if the replacement was successful, false otherwise.
+     */
+    bool replaceOne(std::string pattern, std::string replacement);
+
+    /**
+     * Replaces the random occurrence of a pattern in the input string with a bash script.
+     * @param bashFileName The bash script file name.
+     * @return True if the replacement was successful, false otherwise.
+     */
+    bool replaceWithBashScript(std::string bashFileName);
+
+private:
+    /**
+     * Runs a bash command.
+     * @param bashCommand The bash command to execute.
+     * @return True if the bash command was successful, false otherwise.
+     */
+    bool runBash(const std::string& bashCommand);
+    std::string createTempFile(const std::string& content);
+    std::string readFile(const std::string& fileName);
+};
+
+#endif // MUTATOR_USING_BASH_H
\ No newline at end of file
diff --git a/mutators/lib/CorrectUndefinedBehavior.cpp b/mutators/lib/CorrectUndefinedBehavior.cpp
index 38444930..3db84709 100644
--- a/mutators/lib/CorrectUndefinedBehavior.cpp
+++ b/mutators/lib/CorrectUndefinedBehavior.cpp
@@ -809,5 +809,5 @@ public:
   bool post(llvm::raw_ostream &os) override { return cforge_injector.post(os); }
 };
 
-static RegisterMutator<CorrectUndefinedBehavior>
-    X("CorrectUndefinedBehavior", "correct static undefined behavior");
+// static RegisterMutator<CorrectUndefinedBehavior>
+//     X("CorrectUndefinedBehavior", "correct static undefined behavior");
diff --git a/mutators/lib/Mutator.cpp b/mutators/lib/Mutator.cpp
index 32a068a3..a0947045 100644
--- a/mutators/lib/Mutator.cpp
+++ b/mutators/lib/Mutator.cpp
@@ -24,6 +24,8 @@ clang::CompilerInstance &Mutator::getCompilerInstance() const {
   return getManager().getCompilerInstance();
 }
 
+unsigned Mutator::getSeed() const { return getManager().getSeed(); }
+
 MutationRange Mutator::getMutationRange() const {
   return getManager().getMutationRange();
 }
diff --git a/mutators/lib/MutatorManager.cpp b/mutators/lib/MutatorManager.cpp
index 0a9d2a50..229ab294 100644
--- a/mutators/lib/MutatorManager.cpp
+++ b/mutators/lib/MutatorManager.cpp
@@ -30,6 +30,7 @@ MutatorManager *MutatorManager::getInstance() {
 }
 
 void MutatorManager::srand(unsigned seed) {
+  this->SEED = seed;
   rndgen = std::make_unique<std::mt19937>(seed);
 }
 
@@ -60,8 +61,10 @@ bool MutatorManager::hasMutator(std::string name) {
 
 std::vector<std::string> MutatorManager::getAllMutators() {
   std::vector<std::string> mutators;
-  for (auto &kvpair : getMutatorsMap())
-    mutators.push_back(kvpair.first);
+  for (auto &kvpair : getMutatorsMap()) {
+    if (kvpair.first != "CorrectUndefinedBehavior")
+      mutators.push_back(kvpair.first);
+  }
   return mutators;
 }
 
diff --git a/mutators/lib/MutatorUsingBash.cpp b/mutators/lib/MutatorUsingBash.cpp
new file mode 100644
index 00000000..58f16688
--- /dev/null
+++ b/mutators/lib/MutatorUsingBash.cpp
@@ -0,0 +1,209 @@
+#include <clang/AST/ASTContext.h>
+#include <clang/AST/RecursiveASTVisitor.h>
+#include <clang/Basic/SourceManager.h>
+#include <clang/Lex/Lexer.h>
+#include "MutatorUsingBash.h"
+#include <fstream>
+#include <iostream>
+#include <unistd.h> // For mkstemp
+
+using namespace clang;
+
+std::string MutatorUsingBash::createTempFile(const std::string& content) {
+    // Create a template for mkstemp (must end with "XXXXXX")
+    char tempFileName[] = "/tmp/tempfileXXXXXX";
+
+    // mkstemp generates a unique filename and opens the file securely
+    int fd = mkstemp(tempFileName);
+
+    if (fd == -1) {
+        llvm::outs() << "Failed to create a temporary file.\n";
+        return "";
+    }
+
+    // llvm::outs() << "Temporary file created: " << tempFileName << "\n";
+
+    // Write to the file
+    std::ofstream tempFile(tempFileName);
+    if (tempFile.is_open()) {
+        tempFile << content;
+        tempFile.close();
+    } else {
+        llvm::outs() << "Failed to open temporary file.\n";
+    }
+
+    // Close the file descriptor
+    close(fd);
+
+    return std::string(tempFileName);
+}
+
+std::string MutatorUsingBash::readFile(const std::string& fileName) {
+    std::ifstream file(fileName, std::ios::in | std::ios::binary);
+    if (!file) {
+        llvm::outs() << "Failed to open file: " << fileName << "\n";
+        return "";
+    }
+
+    // Read the entire file content into a string
+    std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+    return content;
+}
+
+bool MutatorUsingBash::replaceAll(std::string pattern, std::string replacement) {
+    SourceManager *SM = &getASTContext().getSourceManager();
+
+    auto MainFileID = SM->getMainFileID();
+    llvm::StringRef FileData = SM->getBufferData(MainFileID);
+
+    std::string EntireFileContent = FileData.str();
+
+    // Create a temporary file
+    std::string tempFileName = createTempFile(EntireFileContent);
+
+    // Run bash command on the string
+    std::string bashCommand = "sed -i -E \'s/" + pattern + "/" + replacement + "/\' \"" + tempFileName + "\" 2>/dev/null";
+    // llvm::outs() << "bashCommand: " << bashCommand << "\n";
+
+    bool result = runBash(bashCommand);
+    if (result == false) {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    std::string output = readFile(tempFileName);
+    if (output == "") {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    // Remove the temporary file
+    std::remove(tempFileName.c_str());
+
+    // Get the start and end SourceLocations of the file
+    SourceLocation StartLoc = SM->getLocForStartOfFile(MainFileID);
+    SourceLocation EndLoc = SM->getLocForEndOfFile(MainFileID);
+
+    // Construct a SourceRange
+    SourceRange FileRange(StartLoc, EndLoc);
+
+    getRewriter().ReplaceText(FileRange, output);
+    return true;
+}
+
+bool MutatorUsingBash::replaceOne(std::string pattern, std::string replacement) {
+    SourceManager *SM = &getASTContext().getSourceManager();
+
+    auto MainFileID = SM->getMainFileID();
+    llvm::StringRef FileData = SM->getBufferData(MainFileID);
+
+    std::string EntireFileContent = FileData.str();
+
+    // Create a temporary file
+    std::string tempFileName = createTempFile(EntireFileContent);
+
+    // Run bash command on the string
+    std::string bashCommand = "bash ~/MetaMut/mutators/scripts/replaceOne.sh " + tempFileName + " \'" + pattern + "\' \'" + replacement + "\' " + std::to_string(getSeed()) + " 2>/dev/null";
+    
+    // llvm::outs() << "bashCommand: " << bashCommand << "\n";
+    bool result = runBash(bashCommand);
+    if (result == false) {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    std::string output = readFile(tempFileName);
+    if (output == "") {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+    
+    // Remove the temporary file
+    std::remove(tempFileName.c_str());
+
+    // Get the start and end SourceLocations of the file
+    SourceLocation StartLoc = SM->getLocForStartOfFile(MainFileID);
+    SourceLocation EndLoc = SM->getLocForEndOfFile(MainFileID);
+
+    // Construct a SourceRange
+    SourceRange FileRange(StartLoc, EndLoc);
+
+    getRewriter().ReplaceText(FileRange, output);
+    return true;
+}
+
+bool MutatorUsingBash::replaceWithBashScript(std::string bashFileName) {
+    SourceManager *SM = &getASTContext().getSourceManager();
+
+    auto MainFileID = SM->getMainFileID();
+    llvm::StringRef FileData = SM->getBufferData(MainFileID);
+
+    std::string EntireFileContent = FileData.str();
+
+    // Create a temporary file
+    std::string tempFileName = createTempFile(EntireFileContent);
+
+    // Run bash command on the string
+    std::string bashCommand = "bash " + bashFileName + " " + tempFileName + " " + std::to_string(getSeed()) +" 2>/dev/null";
+    // llvm::outs() << "bashCommand: " << bashCommand << "\n";
+    bool result = runBash(bashCommand);
+    if (result == false) {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    std::string output = readFile(tempFileName);
+    if (output == "") {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+    
+    // Remove the temporary file
+    std::remove(tempFileName.c_str());
+
+    // Get the start and end SourceLocations of the file
+    SourceLocation StartLoc = SM->getLocForStartOfFile(MainFileID);
+    SourceLocation EndLoc = SM->getLocForEndOfFile(MainFileID);
+
+    // Construct a SourceRange
+    SourceRange FileRange(StartLoc, EndLoc);
+
+    getRewriter().ReplaceText(FileRange, output);
+    return true;
+}
+
+bool MutatorUsingBash::runBash(const std::string& bashCommand) {
+    
+    // Open a pipe to read the output
+    std::array<char, 128> buffer;
+    std::string result;
+    FILE* pipe = popen(bashCommand.c_str(), "r");
+
+    if (!pipe) {
+        std::cerr << "Failed to run bash command." << std::endl;
+        return false;
+    }
+
+    // Read the output from the pipe
+    while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
+        result += buffer.data();
+    }
+
+    // Get the exit code of the command
+    int exitCode = pclose(pipe);
+
+    // Check if the exit code is non-zero
+    if (WIFEXITED(exitCode) && WEXITSTATUS(exitCode) != 0) {
+        std::cerr << "Command failed with exit code: " << WEXITSTATUS(exitCode) << std::endl;
+        return false;
+    }
+
+    // Remove trailing newline from the result
+    if (!result.empty() && result.back() == '\n') {
+        result.pop_back();
+    }
+
+    // llvm::outs() << "bash result: " << result << "\n";
+
+    return true;
+}
\ No newline at end of file
diff --git a/mutators/lib/examples/ChangeBinop.cpp b/mutators/lib/examples/ChangeBinop.cpp
index 9eaf6366..8b6d7c98 100644
--- a/mutators/lib/examples/ChangeBinop.cpp
+++ b/mutators/lib/examples/ChangeBinop.cpp
@@ -63,5 +63,5 @@ private:
   }
 };
 
-static RegisterMutator<ChangeBinop> M(
-    "ChangeBinop", "Change binary expression's operator.");
+// static RegisterMutator<ChangeBinop> M(
+//     "ChangeBinop", "Change binary expression's operator.");
diff --git a/mutators/lib/examples/ChangeDeclRef.cpp b/mutators/lib/examples/ChangeDeclRef.cpp
index 87a40a91..6375c968 100644
--- a/mutators/lib/examples/ChangeDeclRef.cpp
+++ b/mutators/lib/examples/ChangeDeclRef.cpp
@@ -73,5 +73,5 @@ private:
   std::vector<std::vector<clang::VarDecl *>> TheVars;
 };
 
-static RegisterMutator<ChangeDeclRef> M(
-    "ChangeDeclRef", "Change a DeclRefExpr's referenced variable.");
+// static RegisterMutator<ChangeDeclRef> M(
+//     "ChangeDeclRef", "Change a DeclRefExpr's referenced variable.");
diff --git a/mutators/main.cpp b/mutators/main.cpp
index 25e5ea31..e4ab3d7c 100644
--- a/mutators/main.cpp
+++ b/mutators/main.cpp
@@ -38,6 +38,10 @@ cl::opt<bool> listMutators("list-mutators", cl::desc("List all mutators"),
 cl::opt<bool> syntaxOnly("syntax-only", cl::desc("Perform syntax checks only"),
                          cl::init(false), cl::cat(mutOpts));
 
+cl::opt<bool> printMutationState("print-mutations",
+    cl::desc("Print each mutation's results"), cl::init(false),
+    cl::cat(mutOpts));
+
 cl::opt<std::string>
     tryMutators("try-mutators",
                 cl::desc("Try all specific mutators in specified order"),
@@ -61,33 +65,63 @@ std::vector<std::string> splitString(const std::string &str, char delimiter) {
 
 bool tryMutationWithMutators(MutatorManager *manager,
                              const std::vector<std::string> &mutators) {
-  std::string text;
-  llvm::raw_string_ostream oss(text);
-  if (mutators.size() > 1) {
-    manager->setOutStream(oss);
+  json state = json::array();
+  state.push_back(
+      json::object({{"entry", "record"}, {"seed", unsigned(opt::seed)}, {"input", std::string(opt::input)}}));
+  
+  // read file contents
+  std::ifstream file(std::string(opt::input));
+  std::stringstream buffer;
+
+  if (file) {
+    buffer << file.rdbuf();
+  } else {
+    return false;
   }
 
+  std::string buffer_contents = buffer.str();
+  llvm::StringRef srctext = buffer_contents;
+
+  std::string text;
+
   for (auto &m : mutators) {
     if (!manager->hasMutator(m)) {
       llvm::errs() << "nonexistent mutator " << m << "\n";
       continue;
     }
+    text.clear();
+    llvm::raw_string_ostream oss(text);
+    manager->setOutStream(oss);
     manager->setMutator(m);
-    if (manager->mutate()) {
+    bool success = manager->mutate();
+    oss.flush();
+    llvm::StringRef output = llvm::StringRef(text);
+    bool changed = (output.size() != 0) && (output != srctext);
+    if (success && changed) {
       llvm::errs() << "successfully with " << m << "\n";
-      if (mutators.size() > 1) {
+      //if (mutators.size() > 1) {
         if (opt::output == "-")
           llvm::outs() << oss.str();
         else {
+          state.push_back(
+            json::object({{"entry", "mutation"}, {"mutator", m}}));
+
           std::ofstream outfile;
           outfile.open(std::string(opt::output));
           outfile << text;
           outfile.close();
         }
+      //}
+      if (opt::printMutationState) {
+        std::error_code ec;
+        llvm::raw_fd_ostream ofs(std::string(opt::output) + ".json", ec);
+            ofs << state.dump(2) << "\n";
+        ofs.close();
       }
       return true;
-    } else if (oss.str().size())
-      llvm::errs() << "error with " << m << ":\n" << oss.str() << "\n";
+    }
+    // else if (oss.str().size())
+    //   llvm::errs() << "error with " << m << ":\n" << oss.str() << "\n";
   }
   return false;
 }
@@ -117,7 +151,7 @@ int main(int argc, char **argv) {
 
   range.filename = opt::input;
 
-  manager->setOutputFile(opt::output);
+  // manager->setOutputFile(opt::output);
   manager->setSourceFile(opt::input);
   manager->setMutationRange(range);
 
diff --git a/mutators/scripts/replaceOne.sh b/mutators/scripts/replaceOne.sh
new file mode 100644
index 00000000..a76a6464
--- /dev/null
+++ b/mutators/scripts/replaceOne.sh
@@ -0,0 +1,39 @@
+#!/bin/bash
+
+# Check for correct number of arguments
+if [ $# -ne 4 ]; then
+  echo "Usage: $0 <file> <pattern> <replacement> <seed>"
+  exit 1
+fi
+
+FILE=$1
+PATTERN=$2
+REPLACEMENT=$3
+SEED=$4
+
+# Check if the file exists
+if [ ! -f "$FILE" ]; then
+  echo "Error: File '$FILE' not found."
+  exit 1
+fi
+
+# Find all matching lines containing the pattern using grep (no -P option)
+matches=$(grep -nE "$PATTERN" "$FILE" | awk -v seed="$SEED" 'BEGIN {srand(seed); line=""} {if (rand() <= 1/NR) line=$0} END {print line}')
+
+# echo $matches
+
+if [ -z "$matches" ]; then
+  echo "No matches found for the pattern '$PATTERN'."
+  exit 1
+fi
+
+# Extract the line number from the match
+line_number=$(echo "$matches" | cut -d: -f1)
+
+# Use sed to replace the first occurrence on the randomly selected line
+# Obs. the extra argument "" is necessary only in MacOS. It is unnecessary (and not used) on Linux
+sed -Ei "${line_number}s/$PATTERN/$REPLACEMENT/" $FILE
+
+# echo "Replaced a random occurrence of '$PATTERN' in line $line_number with '$REPLACEMENT'."
+
+# cat $FILE
\ No newline at end of file
diff --git a/setup.sh b/setup.sh
index 8ea094e5..cfba154b 100644
--- a/setup.sh
+++ b/setup.sh
@@ -6,13 +6,20 @@ set -e
 # configurations
 : "${CC_DWNDIR:=$(pwd)/compilers}"
 : "${CC_OBJDIR:=$(pwd)/objects}"
-: "${JOBS:=4}"
+: "${JOBS:=8}"
+
+LLVM_CONFIG=$(which llvm-config)
+
+if ! [ -x $LLVM_CONFIG ]; then
+  echo -e "\033[1;31mCannot find an executable llvm-config"
+  exit 1
+fi
 
 # other variables
 PROJ_HOME=$(pwd)
 CFORGE_OBJDIR=${PROJ_HOME}/output
 INSTRUM_PATH=$PROJ_HOME/instrumenter/output/afl-bins
-LLVM_VERSION=$(llvm-config --version | grep -oE '([0-9]+)\.[0-9]+\.[0-9]+' | awk -F. '{print $1}')
+LLVM_VERSION=$($LLVM_CONFIG --version | grep -oE '([0-9]+)\.[0-9]+\.[0-9]+' | awk -F. '{print $1}')
 
 mkdir -p $CC_DWNDIR
 mkdir -p $CC_OBJDIR
@@ -22,13 +29,16 @@ mkdir -p $CFORGE_OBJDIR
 cd $PROJ_HOME/instrumenter
 mkdir -p output; cd output
 cmake .. -DCMAKE_C_COMPILER=gcc-11 -DCMAKE_CXX_COMPILER=g++-11
-make -j4
+make -j8
 
 # download gcc source
 if ! [ -d $CC_DWNDIR/gcc-project-trunk ]; then
   cd $CC_DWNDIR
   echo "[*] downloading gcc's source code"
-  git clone --depth=1 git://gcc.gnu.org/git/gcc.git gcc-project-trunk
+  git clone --depth 1 https://github.com/gcc-mirror/gcc.git gcc-project-trunk
+  cd gcc-project-trunk
+  git fetch --depth 1 origin 87492fb3fd5e7510983e0275a38ba95769335018
+  git checkout 87492fb3fd5e7510983e0275a38ba95769335018
 fi
 
 # gcc-project-trunk.obj
@@ -76,7 +86,10 @@ fi
 if ! [ -d $CC_DWNDIR/llvm-project-trunk ]; then
   cd $CC_DWNDIR
   echo "[*] downloading llvm's source code"
-  git clone --depth=1 git@github.com:llvm/llvm-project llvm-project-trunk
+  git clone --depth 1 https://github.com/llvm/llvm-project.git llvm-project-trunk
+  cd llvm-project-trunk
+  git fetch --depth 1 origin 9bdf683ba6cd9ad07667513d264a2bc02d969186
+  git checkout 9bdf683ba6cd9ad07667513d264a2bc02d969186
 fi
 
 # llvm-project-trunk.obj
diff --git a/test_run.sh b/test_run.sh
new file mode 100644
index 00000000..8469100b
--- /dev/null
+++ b/test_run.sh
@@ -0,0 +1,10 @@
+#!/bin/bash
+
+working_dir=$(pwd)
+
+mkdir workspace
+
+cd workspace
+
+python3 ../fuzzer/run.py -j 10 --wdir $working_dir/workspace --repeat-times 5 --duration 300 --seeds-dir $working_dir/seeds/ --seed 0
+
