diff --git a/mutators/bhis/remove.variable.declaration-59944.cpp b/mutators/bhis/remove.variable.declaration-59944.cpp
new file mode 100644
index 00000000..29bb75b9
--- /dev/null
+++ b/mutators/bhis/remove.variable.declaration-59944.cpp
@@ -0,0 +1,25 @@
+#include <clang/AST/ASTContext.h>
+#include <clang/AST/RecursiveASTVisitor.h>
+#include <clang/Basic/SourceManager.h>
+#include <clang/Lex/Lexer.h>
+#include "Mutator.h"
+#include "MutatorUsingBash.h"
+#include "MutatorManager.h"
+#include <iostream>
+
+using namespace clang;
+
+class RemoveVariableDeclaration : public MutatorUsingBash, public RecursiveASTVisitor<RemoveVariableDeclaration> {
+public:
+    using MutatorUsingBash::MutatorUsingBash;
+    
+    bool mutate() override {
+        return replaceWithBashScript("../scripts/remove.variable.declaration-59944.sh");
+    }
+
+private:
+};
+
+static RegisterMutator<RemoveVariableDeclaration> M(
+    "remove.variable.declaration",
+    "Removes variable declaration that the variable was used for address of operator.");
diff --git a/mutators/include/MutatorUsingBash.h b/mutators/include/MutatorUsingBash.h
new file mode 100644
index 00000000..90a3b10a
--- /dev/null
+++ b/mutators/include/MutatorUsingBash.h
@@ -0,0 +1,58 @@
+#ifndef MUTATOR_USING_BASH_H
+#define MUTATOR_USING_BASH_H
+
+#include <clang/AST/ASTContext.h>
+#include <clang/AST/RecursiveASTVisitor.h>
+#include <clang/Basic/SourceManager.h>
+#include <clang/Lex/Lexer.h>
+#include <string>
+#include <array>
+#include <iostream>
+#include "Mutator.h"
+
+using namespace clang;
+
+class MutatorUsingBash : public Mutator {
+public:
+    using Mutator::Mutator;
+
+    MutatorUsingBash(const char* name, const char* desc)
+        : Mutator(name, desc) {
+        // Additional initialization (if needed)
+    }
+
+    /**
+     * Replaces all occurrences of a pattern in the input string with a replacement string.
+     * @param pattern The pattern to replace.
+     * @param replacement The replacement string.
+     * @return True if the replacement was successful, false otherwise.
+     */
+    bool replaceAll(std::string pattern, std::string replacement);
+
+    /**
+     * Replaces the random occurrence of a pattern in the input string with a replacement string.
+     * @param pattern The pattern to replace.
+     * @param replacement The replacement string.
+     * @return True if the replacement was successful, false otherwise.
+     */
+    bool replaceOne(std::string pattern, std::string replacement);
+
+    /**
+     * Replaces the random occurrence of a pattern in the input string with a bash script.
+     * @param bashFileName The bash script file name.
+     * @return True if the replacement was successful, false otherwise.
+     */
+    bool replaceWithBashScript(std::string bashFileName);
+
+private:
+    /**
+     * Runs a bash command.
+     * @param bashCommand The bash command to execute.
+     * @return True if the bash command was successful, false otherwise.
+     */
+    bool runBash(const std::string& bashCommand);
+    std::string createTempFile(const std::string& content);
+    std::string readFile(const std::string& fileName);
+};
+
+#endif // MUTATOR_USING_BASH_H
\ No newline at end of file
diff --git a/mutators/lib/MutatorUsingBash.cpp b/mutators/lib/MutatorUsingBash.cpp
new file mode 100644
index 00000000..676b7eae
--- /dev/null
+++ b/mutators/lib/MutatorUsingBash.cpp
@@ -0,0 +1,209 @@
+#include <clang/AST/ASTContext.h>
+#include <clang/AST/RecursiveASTVisitor.h>
+#include <clang/Basic/SourceManager.h>
+#include <clang/Lex/Lexer.h>
+#include "MutatorUsingBash.h"
+#include <fstream>
+#include <iostream>
+#include <unistd.h> // For mkstemp
+
+using namespace clang;
+
+std::string MutatorUsingBash::createTempFile(const std::string& content) {
+    // Create a template for mkstemp (must end with "XXXXXX")
+    char tempFileName[] = "/tmp/tempfileXXXXXX";
+
+    // mkstemp generates a unique filename and opens the file securely
+    int fd = mkstemp(tempFileName);
+
+    if (fd == -1) {
+        llvm::outs() << "Failed to create a temporary file.\n";
+        return "";
+    }
+
+    // llvm::outs() << "Temporary file created: " << tempFileName << "\n";
+
+    // Write to the file
+    std::ofstream tempFile(tempFileName);
+    if (tempFile.is_open()) {
+        tempFile << content;
+        tempFile.close();
+    } else {
+        llvm::outs() << "Failed to open temporary file.\n";
+    }
+
+    // Close the file descriptor
+    close(fd);
+
+    return std::string(tempFileName);
+}
+
+std::string MutatorUsingBash::readFile(const std::string& fileName) {
+    std::ifstream file(fileName, std::ios::in | std::ios::binary);
+    if (!file) {
+        llvm::outs() << "Failed to open file: " << fileName << "\n";
+        return "";
+    }
+
+    // Read the entire file content into a string
+    std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
+    return content;
+}
+
+bool MutatorUsingBash::replaceAll(std::string pattern, std::string replacement) {
+    SourceManager *SM = &getASTContext().getSourceManager();
+
+    auto MainFileID = SM->getMainFileID();
+    llvm::StringRef FileData = SM->getBufferData(MainFileID);
+
+    std::string EntireFileContent = FileData.str();
+
+    // Create a temporary file
+    std::string tempFileName = createTempFile(EntireFileContent);
+
+    // Run bash command on the string
+    std::string bashCommand = "sed -i -E \'s/" + pattern + "/" + replacement + "/\' \"" + tempFileName + "\" 2>/dev/null";
+    // llvm::outs() << "bashCommand: " << bashCommand << "\n";
+
+    bool result = runBash(bashCommand);
+    if (result == false) {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    std::string output = readFile(tempFileName);
+    if (output == "") {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    // Remove the temporary file
+    std::remove(tempFileName.c_str());
+
+    // Get the start and end SourceLocations of the file
+    SourceLocation StartLoc = SM->getLocForStartOfFile(MainFileID);
+    SourceLocation EndLoc = SM->getLocForEndOfFile(MainFileID);
+
+    // Construct a SourceRange
+    SourceRange FileRange(StartLoc, EndLoc);
+
+    getRewriter().ReplaceText(FileRange, output);
+    return true;
+}
+
+bool MutatorUsingBash::replaceOne(std::string pattern, std::string replacement) {
+    SourceManager *SM = &getASTContext().getSourceManager();
+
+    auto MainFileID = SM->getMainFileID();
+    llvm::StringRef FileData = SM->getBufferData(MainFileID);
+
+    std::string EntireFileContent = FileData.str();
+
+    // Create a temporary file
+    std::string tempFileName = createTempFile(EntireFileContent);
+
+    // Run bash command on the string
+    std::string bashCommand = "bash ../scripts/replaceOne.sh " + tempFileName + " \'" + pattern + "\' \'" + replacement + "\' " + std::to_string(getSeed()) + " 2>/dev/null";
+    
+    // llvm::outs() << "bashCommand: " << bashCommand << "\n";
+    bool result = runBash(bashCommand);
+    if (result == false) {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    std::string output = readFile(tempFileName);
+    if (output == "") {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+    
+    // Remove the temporary file
+    std::remove(tempFileName.c_str());
+
+    // Get the start and end SourceLocations of the file
+    SourceLocation StartLoc = SM->getLocForStartOfFile(MainFileID);
+    SourceLocation EndLoc = SM->getLocForEndOfFile(MainFileID);
+
+    // Construct a SourceRange
+    SourceRange FileRange(StartLoc, EndLoc);
+
+    getRewriter().ReplaceText(FileRange, output);
+    return true;
+}
+
+bool MutatorUsingBash::replaceWithBashScript(std::string bashFileName) {
+    SourceManager *SM = &getASTContext().getSourceManager();
+
+    auto MainFileID = SM->getMainFileID();
+    llvm::StringRef FileData = SM->getBufferData(MainFileID);
+
+    std::string EntireFileContent = FileData.str();
+
+    // Create a temporary file
+    std::string tempFileName = createTempFile(EntireFileContent);
+
+    // Run bash command on the string
+    std::string bashCommand = "bash " + bashFileName + " " + tempFileName + " " + std::to_string(getSeed()) +" 2>/dev/null";
+    // llvm::outs() << "bashCommand: " << bashCommand << "\n";
+    bool result = runBash(bashCommand);
+    if (result == false) {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+
+    std::string output = readFile(tempFileName);
+    if (output == "") {
+        std::remove(tempFileName.c_str());
+        return false;
+    }
+    
+    // Remove the temporary file
+    std::remove(tempFileName.c_str());
+
+    // Get the start and end SourceLocations of the file
+    SourceLocation StartLoc = SM->getLocForStartOfFile(MainFileID);
+    SourceLocation EndLoc = SM->getLocForEndOfFile(MainFileID);
+
+    // Construct a SourceRange
+    SourceRange FileRange(StartLoc, EndLoc);
+
+    getRewriter().ReplaceText(FileRange, output);
+    return true;
+}
+
+bool MutatorUsingBash::runBash(const std::string& bashCommand) {
+    
+    // Open a pipe to read the output
+    std::array<char, 128> buffer;
+    std::string result;
+    FILE* pipe = popen(bashCommand.c_str(), "r");
+
+    if (!pipe) {
+        std::cerr << "Failed to run bash command." << std::endl;
+        return false;
+    }
+
+    // Read the output from the pipe
+    while (fgets(buffer.data(), buffer.size(), pipe) != nullptr) {
+        result += buffer.data();
+    }
+
+    // Get the exit code of the command
+    int exitCode = pclose(pipe);
+
+    // Check if the exit code is non-zero
+    if (WIFEXITED(exitCode) && WEXITSTATUS(exitCode) != 0) {
+        std::cerr << "Command failed with exit code: " << WEXITSTATUS(exitCode) << std::endl;
+        return false;
+    }
+
+    // Remove trailing newline from the result
+    if (!result.empty() && result.back() == '\n') {
+        result.pop_back();
+    }
+
+    // llvm::outs() << "bash result: " << result << "\n";
+
+    return true;
+}
\ No newline at end of file
diff --git a/mutators/scripts/remove.variable.declaration-59944.sh b/mutators/scripts/remove.variable.declaration-59944.sh
new file mode 100644
index 00000000..101f42d1
--- /dev/null
+++ b/mutators/scripts/remove.variable.declaration-59944.sh
@@ -0,0 +1,37 @@
+#!/bin/bash
+
+# Check if a file argument is provided
+if [ $# -ne 2 ]; then
+    echo "Usage: $0 <file> <seed>"
+    exit 1
+fi
+
+file="$1"
+SEED=$2
+
+#### Randomly pick one
+# Find all occurrences of address of variable
+matches=$(grep -oE '\&([a-zA-Z_][a-zA-Z0-9_]*)' "$file" | \
+sed -E 's/\&([a-zA-Z_][a-zA-Z0-9_]*)/\1/')
+
+# If no matches are found, exit
+if [ -z "$matches" ]; then
+  echo "No matching patterns found."
+  exit 0
+fi
+
+# Convert matches to an array and randomly pick one
+varible=$(echo "$matches" | awk -v seed="$SEED" 'BEGIN {srand(seed); line=""} {if (rand() <= 1/NR) line=$0} END {print line}')
+
+
+# Find the variable declaration
+matches=$(grep -noE "([a-zA-Z_][a-zA-Z0-9_]* )*[a-zA-Z_][a-zA-Z0-9_]* $varible( = (.*))?;" "$file")
+
+if [ -z "$matches" ]; then
+  echo "No matching patterns found."
+  exit 0
+fi
+
+line=$(echo "$matches" | head -n 1 | cut -d: -f1)
+
+sed -i -E "${line}s/([a-zA-Z_][a-zA-Z0-9_]* )*[a-zA-Z_][a-zA-Z0-9_]* $varible( = (.*))?;//" "$file"
diff --git a/mutators/scripts/replaceOne.sh b/mutators/scripts/replaceOne.sh
new file mode 100644
index 00000000..a76a6464
--- /dev/null
+++ b/mutators/scripts/replaceOne.sh
@@ -0,0 +1,39 @@
+#!/bin/bash
+
+# Check for correct number of arguments
+if [ $# -ne 4 ]; then
+  echo "Usage: $0 <file> <pattern> <replacement> <seed>"
+  exit 1
+fi
+
+FILE=$1
+PATTERN=$2
+REPLACEMENT=$3
+SEED=$4
+
+# Check if the file exists
+if [ ! -f "$FILE" ]; then
+  echo "Error: File '$FILE' not found."
+  exit 1
+fi
+
+# Find all matching lines containing the pattern using grep (no -P option)
+matches=$(grep -nE "$PATTERN" "$FILE" | awk -v seed="$SEED" 'BEGIN {srand(seed); line=""} {if (rand() <= 1/NR) line=$0} END {print line}')
+
+# echo $matches
+
+if [ -z "$matches" ]; then
+  echo "No matches found for the pattern '$PATTERN'."
+  exit 1
+fi
+
+# Extract the line number from the match
+line_number=$(echo "$matches" | cut -d: -f1)
+
+# Use sed to replace the first occurrence on the randomly selected line
+# Obs. the extra argument "" is necessary only in MacOS. It is unnecessary (and not used) on Linux
+sed -Ei "${line_number}s/$PATTERN/$REPLACEMENT/" $FILE
+
+# echo "Replaced a random occurrence of '$PATTERN' in line $line_number with '$REPLACEMENT'."
+
+# cat $FILE
\ No newline at end of file
